lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -183,13 +612,58 @@ class ConnectorSourceNotify[In: Any val]
       end
     end
 
-  fun ref accepted(source: ConnectorSource[In] ref) =>
+  //////////////////////////////
+  // Connection State Management
+  //////////////////////////////
+  fun ref accepted(source: ConnectorSource[In] ref, session_id: RoutingId) =>
     @printf[I32]((_source_name + ": accepted a connection\n").cstring())
+    if _fsm_state isnt _ProtoFsmDisconnected then
+      @printf[I32]("ERROR: %s.connected: state is %d\n".cstring(),
+        __loc.type_name().cstring(), _fsm_state())
+      Fail()
+    end
+    // reset/clear local state
+    _fsm_state = _ProtoFsmConnected
     _header = true
+    _session_active = true
+    _session_id = session_id
+    _clear_streams()
+    _credits = _max_credits
+    _prep_for_rollback = false
     source.expect(_header_size)
 
   fun ref closed(source: ConnectorSource[In] ref) =>
-    @printf[I32]("ConnectorSource connection closed\n".cstring())
+    @printf[I32]("ConnectorSource connection closed 0x%lx\n".cstring(), source)
+    _session_active = false
+    _fsm_state = _ProtoFsmDisconnected
+    // TODO [post-source-migration] When john's work is ready, this assumption
```

Comment:
What work do you mean?  Can you replace "John's work" with a description of the property that will be added?